substrate-node/pallets/pallet-tft-price/src/lib.rs

```diff
@@ -59,119 +73,160 @@ pub mod crypto {
+    pub trait Config: frame_system::Config + CreateSignedTransaction<Call<Self>> {
+        // type Event: From<Event<Self>> + Into<<Self as system::Config>::Event>;
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        // Add other types and constants required to configure this pallet.
+        type AuthorityId: AppCrypto<Self::Public, Self::Signature>;
+        type Call: From<Call<Self>>;
+        /// Origin for restricted extrinsics
+        /// Can be the root or another origin configured in the runtime
+        type RestrictedOrigin: EnsureOrigin<Self::Origin>;
     }
 
-    add_extra_genesis {
-        config(allowed_origin): Option<T::AccountId>;
-
-        build(|_config| {
-            AllowedOrigin::<T>::set(_config.allowed_origin.clone());
-        });
-    }
-}
-
-decl_event! {
-    pub enum Event<T> where AccountId = <T as frame_system::Config>::AccountId {
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    // pub enum Event<T> where AccountId = <T as frame_system::Config>::AccountId {
+    pub enum Event<T: Config> {
         PriceStored(u32),
-        OffchainWorkerExecuted(AccountId),
+        OffchainWorkerExecuted(T::AccountId),
         AveragePriceStored(u32),
     }
-}
 
-decl_error! {
-    pub enum Error for Module<T: Config> {
+    #[pallet::error]
+    // pub enum Error for Module<T: Config> {
+    pub enum Error<T> {
         ErrFetchingPrice,
         OffchainSignedTxError,
         NoLocalAcctForSigning,
         AccountUnauthorizedToSetPrice,
     }
-}
 
-decl_module! {
-    pub struct Module<T: Config> for enum Call where origin: T::Origin {
-        type Error = Error<T>;
-
-        fn deposit_event() = default;
-
-        #[weight = 100_000_000 + T::DbWeight::get().writes(3)]
-        pub fn set_prices(origin, price: u32, block_number: T::BlockNumber) -> DispatchResultWithPostInfo {
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn tft_price)]
+    pub type TftPrice<T> = StorageValue<_, u32, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn last_block_set)]
+    pub type LastBlockSet<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn average_tft_price)]
+    pub type AverageTftPrice<T> = StorageValue<_, u32, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn get_value)]
+    pub type TftPriceHistory<T> = StorageMap<_, Blake2_128Concat, BufferIndex, u32, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn range)]
+    pub type BufferRange<T> = StorageValue<_, (BufferIndex, BufferIndex), ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn allowed_origin)]
+    pub type AllowedOrigin<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(0)]
```

Comment:
Set some weight here.

Reply:
the weight is based on what? 
is there a rule or something to be able to evaluate it?

Reply:
base weight: `100_000_000` + number of reads / writes

Reply:
We can benchmark all the extrinsics to get proper weighting, this will need to be done in later version

Reply:
> base weight: `100_000_000` + number of reads / writes

great
I get it